[Collections] Use image cache for selection icons

romoore opened the pull request:
The selected/unselected state icons are reloaded from disk on every
prepareForReuse call.  Using the cache-backed `-imageNamed` family of
methods in UIImage reduces the overhead required to reload those images.

Closes #1637

samnm commented on components/CollectionCells/src/MDCCollectionViewCell.m:
```diff
@@ -424,16 +424,32 @@ - (void)updateInterfaceForEditing {
+                  [UIImage imageNamed:@"ic_check_circle"
+                             inBundle:[MDCCollectionViewCell
+                                          bundleWithName:@"MaterialIcons_ic_check_circle"]
+        compatibleWithTraitCollection:nil];
```
As discussed offline, this should probably be moved in to the MDCIcon+\<Name\> generated by sync_icons.sh.

ianegordon commented on components/CollectionCells/src/MDCCollectionViewCell.m:
```diff
@@ -424,16 +424,32 @@ - (void)updateInterfaceForEditing {
 
 #pragma mark - Selecting
 
++ (NSBundle *)bundleWithName:(NSString *)bundleName {
```
Ideally we expose this in MDCIcons.  I would add a TODO here noting that.

Consider a bit more context on the method name, perhaps:
- bundleWithIconName:(NSString *)iconName

Naming is hard.

ianegordon commented on components/CollectionCells/src/MDCCollectionViewCell.m:
```diff
@@ -424,16 +424,32 @@ - (void)updateInterfaceForEditing {
   if (selected) {
     _editingSelectorImageView.image =
-        [UIImage imageWithContentsOfFile:[MDCIcons pathFor_ic_check_circle]];
+                  [UIImage imageNamed:@"ic_check_circle"
```
formatting.

ianegordon commented on components/CollectionCells/src/MDCCollectionViewCell.m:
```diff
@@ -424,16 +424,32 @@ - (void)updateInterfaceForEditing {
+    [UIImage imageNamed:@"ic_radio_button_unchecked"
+               inBundle:[MDCCollectionViewCell
+                         bundleWithName:@"MaterialIcons_ic_radio_button_unchecked"]
+compatibleWithTraitCollection:nil];
```
Formatting.
Since compatibleWIthTraitCollection: won't fit, align the ':' for the other parameters, then indent this one 4 space past the original "[UIImage imageName:..."

Commit:
[Icons] Add icon method for use with imageNamed

Icon images are currently loaded manually from a path in a bundle and therefore bypass the internal UIImage caching mechanism. This change will add new methods that provide the icon images via the -imageNamed family of methods.

Closes #1637, #1606

Commit:
Fixing the imports in the sync_icons script

Commit:
Fixing arrow_back

Commit:
Using new arrow_back icon

ianegordon commented on components/private/Icons/src/MDCIcons.h:
```diff
@@ -20,7 +20,8 @@
 
  Individual icons can be accessed by importing their MaterialIcons+<icon_name>.h header and calling
- [MDCIcons pathFor_<icon_name>] to get the icon's file system path.
+ [MDCIcons pathFor_<icon_name>] to get the icon's file system path or calling
```
s/pathor/path or/

romoore replied:
I don't see the "pathor" here, I don't think this file had the paste bug.

ianegordon commented on components/private/Icons/src/MDCIcons+BundleLoader.h:
```diff
@@ -20,12 +20,15 @@
 
  Individual icons can be accessed by importing their MaterialIcons+<icon_name>.h header and calling
- [MDCIcons pathFor_<icon_name>] to get the icon's file system path.
+ [MDCIcons pathFor_<icon_name>] to get the icon's file system pathor calling
```
s/pathor/path or/

romoore replied:
Done

ianegordon commented on scripts/sync_icons.sh:
```diff
@@ -230,6 +236,12 @@ __attribute__((visibility("default")))
++ (nullable UIImage *)imageFor_$icon_name {
+  NSBundle *bundle = [self bundleNamed:kBundleName];
+  return [UIImage imageNamed:kIconName inBundle:bundle
+             compatibleWithTraitCollection:nil];
```
If you line break in the middle of a method call, all parameters must be on their own line.
Or, to put it another way:
```
return [UIImage imageNamed:kIconName
                  inBundle:bundle
    compatibleWithTraitCollection:nil];
```

Note that compatible won't fit if you try and align colons, so it is indented 4 spaces.

romoore replied:
Done

Commit:
Formatting/spelling fixes

Commit:
More cleanup, fixing script typo for documentation

ianegordon commented:
Waiting for Travis, but local testing was successful.
ic_arrow_back continued to have the manually added switch between the old and new style as expected.
LGTM.

samnm commented on components/private/Icons/icons/ic_check/src/MaterialIcons+ic_check.h:
```diff
@@ -29,4 +29,10 @@
+ Returns the image for the ic_check image contained in
+ MaterialIcons_ic_check.bundle.
+ */
++ (nullable UIImage *)imageFor_ic_check;
```
These files are autogenerated by sync_icons.sh. Did you look in to updating sync_icons with the imageFor method?

samnm replied:
Oh, you did. LGTM.

Commit:
Adding unit tests for new methods
